### ISSUE TITLE
Expressions: support boolean/logical operators

### ISSUE THREAD

**Issue by d1nuc0m** [FreeCAD/FreeCAD#8008]
### Is there an existing request for this?

- [X] I have searched the existing requests

### Forums discussion

https://forum.freecadweb.org/viewtopic.php?f=3&t=42418

### Subproject(s) affected?

Expressions

### Idea description

Extend the expression parser supporting the boolean operators (AND/OR/NOT).

At least for AND/OR there is a workaround with nested conditional expressions (see discussion, thanks [Roy_043](https://forum.freecadweb.org/memberlist.php?mode=viewprofile&u=22936) / @Roy-043):
 
```
AND:
(1 < 2 ? 1 : 0) * (2 < 3 ? 1 : 0) == 1 ? 100 : 20

OR:
(1 < 2 ? 1 : 0) + (2 < 3 ? 1 : 0) >= 1 ? 100 : 20
```

but having proper operators would make expressions more readable, something like:

```
AND
(1 < 2 AND 2 < 3) ? 100 : 20
OR
(1 < 2 OR 2 < 3) ? 100 : 20
```

While for the not operator, no workaround comes to my mind.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by Daniel-Khodabakhsh** [comment]
> @Daniel-Khodabakhsh any interest in tckling this ticket ?

Sure!

**Comment by d1nuc0m** [comment]
I don't know if it can be useful, but it looks like RealThunder's fork includes a sort of AND/OR, as AND_OP and OR_OP in [/src/App/ExpressionParser.y](https://github.com/realthunder/FreeCAD/blob/LinkDaily/src/App/ExpressionParser.y)

**Comment by Daniel-Khodabakhsh** [comment]
Oh interesting, thanks @d1nuc0m. @luzpaz with that in mind, what's the best way to add that to FreeCAD? Should I poke RealThunder to merge it back, copy paste a new branch with attribution to RealThunder, or do it more-or-less from scratch? I was thinking it would be neat if we had functions (`and(...)`, `or(...)`) as well as operators (`&&`, `||`, `and`. `or`).

**Comment by luzpaz** [comment]
@chennes care to weigh-in on how to proceed ?

**Comment by d1nuc0m** [comment]
> @chennes care to weigh-in on how to proceed ?

@chennes just a ping reminder

**Comment by app4soft** [comment]
> While for the NOT operator, no workaround comes to my mind.

For NOT there is already `!=` (could be used for both numerical and text strings).

And here is workaround formula to compare 2 items and return 1 of 3 possible answers:

```
=((A > B ? 2:3)+(A < B ? 2:1))!=4?(A > B ? <<A is bigger>>:<<B is bigger>>):<<A equal to B>>
```

**Comment by maxwxyz** [comment]
@pieterhijma is this interesting for: #12120

**Comment by pieterhijma** [comment]
> @pieterhijma is this interesting for: #12120

No, a potential solution for #12120 would be too much out of scope of this issue.

**Comment by NomAnor** [comment]
This would be useful to conditionally suppress features (#12096).

**Comment by mnesarco** [comment]
Having three functions may be enough for all cases:

```not(v)```, 
```all(a; b; c; ...)```, 
```any(a; b; c; ...)```

I don't know the details of the current parser, but as it already have some functions support, it should be easier to implement new functions than operators because operators have inline syntax and precedence considerations.
